src/Columns/ColumnMap.cpp

```diff
@@ -0,0 +1,395 @@
+#include <Columns/ColumnMap.h>
+#include <Columns/IColumnImpl.h>
+#include <DataStreams/ColumnGathererStream.h>
+#include <IO/WriteBufferFromString.h>
+#include <IO/Operators.h>
+#include <ext/map.h>
+#include <ext/range.h>
+#include <Common/typeid_cast.h>
+#include <Common/assert_cast.h>
+#include <Common/WeakHash.h>
+#include <Core/Field.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int NOT_IMPLEMENTED;
+    extern const int CANNOT_INSERT_VALUE_OF_DIFFERENT_SIZE_INTO_TUPLE;
+    extern const int LOGICAL_ERROR;
+}
+
+
+std::string ColumnMap::getName() const
+{
+    WriteBufferFromOwnString res;
+    res << "Map(" << columns[0]->getName() << ", " << columns[1]->getName() << ")";
+    return res.str();
+}
+
+ColumnMap::ColumnMap(MutableColumns && mutable_columns)
+{
+    columns.reserve(mutable_columns.size());
+    for (auto & column : mutable_columns)
```

Comment:
Here you can init `ColumnMap::columns` to have less than 2 columns, that would cause a crash at run-time in say `ColumnMap::cloneEmpty()` or any other method that assumes that there are exactly 2 columns.

Reply:
Consider checking that each of `mutable_columns` is actually a `ColumnArray`.